=== frame.c
==================================================================
--- frame.c	(revision 52633)
+++ frame.c	(local)
@@ -44,7 +44,9 @@
 static
 int valid_idchar(char c)
 {
-  return (c >= 'A' && c <= 'Z') || (c >= '0' && c <= '9');
+  // iTunes sometimes writes v2.3/v2.4 tags in v2.2 format
+  // with a space, so we need to allow this.
+  return (c >= 'A' && c <= 'Z') || (c >= '0' && c <= '9') || c == ' ';
 }
 
 /*
