=== utf16.c
==================================================================
--- utf16.c	(revision 59423)
+++ utf16.c	(local)
@@ -277,6 +277,12 @@
       break;
     }
   }
+  
+  /* Bug 14728
+    If there is no BOM, assume LE, this is what appears in the wild -andy
+  */
+  if (byteorder == ID3_UTF16_BYTEORDER_ANY)
+    byteorder = ID3_UTF16_BYTEORDER_LE;
 
   utf16ptr = utf16;
   while (end - *ptr > 0 && (*utf16ptr = id3_utf16_get(ptr, byteorder)))
